crates/matrix-sdk/src/room/common.rs

```diff
@@ -817,6 +819,65 @@ impl Common {
         let request = delete_tag::v3::Request::new(&user_id, self.inner.room_id(), tag.as_ref());
         self.client.send(request, None).await
     }
+
+    /// Sets whether this room is a DM.
+    ///
+    /// When setting this room as DM, it will be marked as DM for all active
+    /// members of the room. When unsetting this room as DM, it will be
+    /// unmarked as DM for all users, not just the members.
+    ///
+    /// Returns the [`set_global_account_data::v3::Response`] from the server.
+    ///
+    /// # Arguments
+    /// * `is_direct` - Whether to mark this room as direct.
+    pub async fn set_is_direct(&self, is_direct: bool) -> Result<()> {
+        let user_id = self
+            .client
+            .user_id()
+            .await
+            .ok_or_else(|| Error::from(HttpError::AuthenticationRequired))?;
+
+        let content = self
+            .client
+            .store()
+            .get_account_data_event(GlobalAccountDataEventType::Direct)
+            .await?
+            .map(|e| e.deserialize_as::<DirectEvent>())
+            .transpose()?
+            .map(|e| e.content);
+
+        let mut content = if let Some(content) = content {
+            content
+        } else if is_direct {
+            ruma::events::direct::DirectEventContent(BTreeMap::new())
+        } else {
+            return Ok(());
```

Comment:
This believes that just because we do not have the event in our local store (which we should consider more of a cache instead), the room is _globally_ not marked as a DM. While this might hold water under current sync for most cases, at least once we have sliding_sync coming in, we can not assume anymore that we've seen all these potential events. But even today, if we are looking at a stale state (right after having started and not synced up again), we might have not seen the event yet, the server already knows about. Stopping early here, sounds like a risky move IMHO. I'd rather we continue in this scenario by sending the event.

Reply:
I removed the early return. Wouldn't overriding the state be still an issue?